psycopg/psycopg/types/numpy.py

```diff
@@ -0,0 +1,153 @@
+"""
+Adapters for numpy types.
+"""
+
+# Copyright (C) 2022 The Psycopg Team
+
+import operator
+from typing import Any, Optional, cast
+
+from psycopg.types.numeric import (
+    dump_int_to_numeric_binary,
+)
+
+import numpy as np
+
+from .. import postgres
+from .._struct import pack_int2, pack_int4, pack_int8
+from ..abc import AdaptContext, Buffer
+from ..adapt import Dumper
+from ..pq import Format
+
+
+class _NPIntDumper(Dumper):
+    def dump(self, obj: Any) -> bytes:
+        return str(obj).encode()
+
+    def quote(self, obj: Any) -> bytes:
+        value = self.dump(obj)
+        return value if obj >= 0 else b" " + value
+
+
+class NPInt2Dumper(_NPIntDumper):
+    oid = postgres.types["int2"].oid
+
+
+class NPInt4Dumper(_NPIntDumper):
+    oid = postgres.types["int4"].oid
+
+
+class NPInt8Dumper(_NPIntDumper):
+    oid = postgres.types["int8"].oid
+
+
+class NPBooleanDumper(_NPIntDumper):
+    oid = postgres.types["bool"].oid
+
+    def dump(self, obj: np.bool_) -> bytes:
+        return "t".encode() if obj is True else "f".encode()
+
+
+class NPIntNumericDumper(_NPIntDumper):
+    oid = postgres.types["numeric"].oid
+
+
+# Binary Dumpers
+
+
+class NPInt2BinaryDumper(NPInt2Dumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> bytes:
+        return pack_int2(int(obj))
+
+
+class NPInt4BinaryDumper(NPInt4Dumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> bytes:
+        return pack_int4(int(obj))
+
+
+class NPInt8BinaryDumper(NPInt8Dumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> bytes:
+        return pack_int8(int(obj))
+
+
+class NPBooleanBinaryDumper(NPBooleanDumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> bytes:
+        return b"\x01" if obj else b"\x00"
+
+
+# dump_int_to_numeric_binary() doesn't work with numpy types
+# because it they don't implement a `bit_length()` method.
+# `NPUint64Adapter` wraps the numpy type, providing the missing
+# method and supporting all the other arithmetic operation
+# performed inside the function
+class NPUInt64Adapter:
+    _obj: np.uint64
+
+    def __init__(self, obj: np.uint64) -> None:
+        self._obj = obj
+
+    def bit_length(self) -> int:
+        return self._obj.bit_count()
+
+    def __gt__(self, n: int) -> bool:
+        return self._obj > n
+
+    def __lt__(self, n: int) -> bool:
+        return self._obj < n
+
+    def __mod__(self, n: int) -> int:
+        return int(self._obj % n)
+
+    def __rfloordiv__(self, n: int) -> int:
+        return self._obj // n
+
+    def __ifloordiv__(self, n: int) -> int:
+        return int(operator.ifloordiv(self._obj, n))
+
+
+class NPIntBinaryNumericDumper(NPIntNumericDumper):
+
+    format = Format.BINARY
+
+    def dump(self, obj: Any) -> Buffer:
+
+        np_adapted = NPUInt64Adapter(obj)
+
+        # Casting to int mainly to silence mypy complaints
+        return dump_int_to_numeric_binary(cast(int, np_adapted))
```

Comment:
I'd rather not waste opcodes in the dumpers, they are a pretty hot path.

```suggestion
        return dump_int_to_numeric_binary(int, np_adapted)  # type: ignore
```